jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java

```diff
@@ -0,0 +1,540 @@
+
+                return c1.getKey().compareTo(c2.getKey());
+            });
+            sorted.addAll(_cache.values());
+
+            // TODO: Can we remove the buffers from the content before evicting.
+            // Invalidate least recently used first
+            for (CachingHttpContent content : sorted)
+            {
+                if (_cache.size() <= _maxCachedFiles && _cachedSize.get() <= _maxCacheSize)
+                    break;
+                removeFromCache(content);
+            }
+        }
+    }
+
+    protected void removeFromCache(CachingHttpContent content)
+    {
+        CachingHttpContent removed = _cache.remove(content.getKey());
+        if (removed != null)
+        {
+            removed.release();
+            _cachedSize.addAndGet(-removed.getContentLengthValue());
+        }
+    }
+
+    public void flushCache()
+    {
+        for (CachingHttpContent content : _cache.values())
+        {
+            removeFromCache(content);
+        }
+    }
+
+    /**
+     * Tests whether the given HttpContent is cacheable, and if there is enough room to fit it in the cache.
+     *
+     * @param httpContent the HttpContent to test.
+     * @return whether the HttpContent is cacheable.
+     */
+    protected boolean isCacheable(HttpContent httpContent)
+    {
+        if (httpContent == null)
+            return true;
+
+        if (httpContent.getResource().isDirectory())
+            return false;
+
+        if (_maxCachedFiles <= 0)
+            return false;
+
+        // Will it fit in the cache?
+        long len = httpContent.getContentLengthValue();
+        if (len <= 0)
+            return false;
+        if (httpContent instanceof FileMappedHttpContentFactory.FileMappedContent)
+             return true;
+        return (len <= _maxCachedFileSize && len <= _maxCacheSize);
+    }
+
+    @Override
+    public HttpContent getContent(String path) throws IOException
+    {
+        CachingHttpContent cachingHttpContent = _cache.get(path);
+        if (cachingHttpContent != null)
+        {
+            cachingHttpContent.setLastAccessedNanos(NanoTime.now());
+            if (cachingHttpContent.isValid())
+                return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+            else
+                removeFromCache(cachingHttpContent);
+        }
+
+        HttpContent httpContent = _authority.getContent(path);
+        if (isCacheable(httpContent))
+        {
+            AtomicBoolean wasAdded = new AtomicBoolean(false);
+            cachingHttpContent = _cache.computeIfAbsent(path, p ->
+            {
+                wasAdded.set(true);
+                CachingHttpContent cachingContent = (httpContent == null) ? newNotFoundContent(p) : newCachedContent(p, httpContent);
+                _cachedSize.addAndGet(cachingContent.getContentLengthValue());
+                return cachingContent;
+            });
+
+            if (wasAdded.get())
```

Comment:
Is it worthwhile having the `wasAdded` atomic?  Why not just call `shrinkCache` anyway and that will do a volitile lookup of the size, which is the same cost.

Reply:
`shrinkCache` will have to do 2 volatile lookups, `_cache.size()` and `_cachedSize.get()`.

Ideally we wouldn't need the AtomicBoolean here, it is just because we can't know if `computeIfAbsent` created a new entry or not.